docs/src/main/asciidoc/security-customization.adoc

```diff
@@ -620,6 +620,51 @@ are using an executor that is capable of propagating the identity (e.g. no `Comp
 to make sure that Quarkus can propagate it. For more information see the
 xref:context-propagation.adoc[Context Propagation Guide].
 
+[[observe-security-events]]
+== Observe security events
+
+Quarkus fires security events, such as authentication or authorization failure, with the `io.quarkus.security.spi.runtime.SecurityEventProducer` CDI bean.
+Default `SecurityEventProducer` implementation fires security events as CDI events.
+This way, you can define an observer method that will be notified when security event was fired.
+
+[source,java]
+----
+package org.acme.security;
+
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.spi.runtime.AuthenticationEvent;
+import io.quarkus.security.spi.runtime.AuthorizationEvent;
+import io.quarkus.security.spi.runtime.SecurityEvent;
+import io.vertx.ext.web.RoutingContext;
+import jakarta.enterprise.event.Observes;
+import jakarta.enterprise.event.ObservesAsync;
+
+import java.util.Map;
+
+public class SecurityEventObserver {
+
+    void observe(@Observes SecurityEvent event) {                   <1>
+        SecurityIdentity identity = event.getSecurityIdentity();
+    }
+
+    void observe(@Observes AuthenticationEvent event) {             <2>
+        Throwable failure = event.getAuthenticationFailure();
+    }
+
+    void observe(@Observes AuthorizationEvent event) {              <3>
+        RoutingContext ctx = (RoutingContext) event.getEventProperties().get(RoutingContext.class.getName());
+    }
+
+    void observeAsync(@ObservesAsync SecurityEvent event) {         <4>
+        Map<String, Object> eventProperties = event.getEventProperties();
+    }
+}
+----
+<1> This observer will be notified with all the security events.
+<2> The `AuthenticationEvent` event is fired whenever authentication failed.
+<3> The `AuthorizationEvent` event is used for failed security checks and HTTP permission checks.
+<4> Both synchronous and asynchronous CDI observer can be used to observe security events.
```

Comment:
Is using async observers more performant ? If yes then we should recommend using the async

Reply:
> I would recommend to fire both, a sync and async event.

This is performance sensitive feature, likely to run almost on every request if for example we are talking authentication success event. No worry, code I added analyze observers and only fire events based on what what events are observed (you can combine it, you can have both synchronous and asynchronous for same event).

> And yes, there are differences. For example, if a sync observer throws an exception the processing of the event is aborted. If an async observer throws an exception then the processing of the event is NOT aborted.

Thanks.

Reply:
> > I would recommend to fire both, a sync and async event.
> 
> This is performance sensitive feature, likely to run almost on every request if for example we are talking authentication success event. No worry, code I added analyze observers and only fire events based on what what events are observed (you can combine it, you can have both synchronous and asynchronous for same event).

Well, ArC does optimize injected `Event`s so that `Event#fire()` is basically no-op if no matching observer exists. But +1 if you can optimize even further...

Reply:
Noted, and thanks again; I don't want to pre-optimize, this code  https://github.com/quarkusio/quarkus/blob/6d6ad1e10bfa32aa0f6502064b738302e90c8cd4/extensions/security/runtime/src/main/java/io/quarkus/security/runtime/event/CDISecurityEventProducer.java#L63 probably provides small advantage against `io.quarkus.arc.impl.EventImpl`, but the code is short, so I'll keep it.

Reply:
> I would recommend to fire both, a sync and async event. And yes, there are differences. For example, if a sync observer throws an exception the processing of the event is aborted. If an async observer throws an exception then the processing of the event is NOT aborted.
> 
> CC @Ladicek @manovotn

Yes, having both is IMO good idea. The optimization here makes sure it's not fired when not needed and various apps are free to choose whether they want to handle it in a sync or async fasion, so :+1:

Reply:
> > Well, ArC does optimize injected Events so that Event#fire()
> 
> I should probably mention I'm not injecting the event as there can be many types of the events.

You should be able to operate just on `Event<SecurityEvent>`. Any observers using subtype of `SecurityEvent` will still get notified.